proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/keychain/SymmetricKeyProvider.java

```diff
@@ -0,0 +1,396 @@
+                    onGenerateKey.configureBuilder(builder);
+                }
+            }
+        });
+    }
+
+    /**
+     * Create a new instance of this class, configured to generate keys for AES-CBC cipher.
+     * @param keyAlias Key alias that identifies the key in Android KeyStore.
+     * @param keySize Number of bits for generated key.
+     * @param randomizedEncryption Whether encryption using this key must be sufficiently randomized
+     *                             to produce different ciphertexts for the same plaintext every time.
+     * @param onGenerateKey Additional key configuration.
+     * @return Instance of this class or {@code null} in case that Android KeyStore is not available.
+     */
+    @Nullable
+    public static SymmetricKeyProvider getAesCbcKeyProvider(
+            @NonNull final String keyAlias,
+            final int keySize,
+            final boolean randomizedEncryption,
+            @Nullable final OnGenerateKey onGenerateKey) {
+        return getSymmetricKeyProvider(keyAlias, new OnGenerateKey() {
+            @Override
+            public void configureBuilder(@NonNull KeyGenParameterSpec.Builder builder) {
+                builder
+                        .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                        .setRandomizedEncryptionRequired(randomizedEncryption);
+                if (keySize != 0) {
+                    builder.setKeySize(keySize);
+                }
+                if (onGenerateKey != null) {
+                    onGenerateKey.configureBuilder(builder);
+                }
+            }
+        });
+    }
+
+    /**
+     * Get existing secret key or create a new one if such key is not exist in Android KeyStore.
+     * @param context Android context object.
+     * @param forceCreateOnFailure If {@code true} then key is re-generated in case that key exists
+     *                             in KeyStore but cannot be obtained.
+     * @return Instance of {@link SecretKey} obtained from Android KeyStore or {@code null} in case of failure.
+     */
+    @Nullable
+    public SecretKey getOrCreateSecretKey(@NonNull Context context, boolean forceCreateOnFailure) {
+        synchronized (SymmetricKeyProvider.class) {
+            if (hasSecretKey()) {
+                try {
+                    return (SecretKey) keyStore.getKey(keyAlias, null);
+                } catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableKeyException e) {
+                    PA2Log.e("SymmetricKeyProvider: " + keyAlias + "Failed to get key. Exception: " + e.getMessage());
+                    if (!forceCreateOnFailure) {
+                        return null;
+                    }
+                    removeSecretKey();
+                }
+            }
+            final SecretKey newSecretKey;
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
```

Comment:
That's not necessary. StrongBox is just a next level of protection. There are devices secure enough (e.g. keys are hardware backed), but the security is not implemented in StrongBox. So, enforcing StrongBox makes no sense, because only few phones manufactured by Google, supports this feature. See [article for more details](https://proandroiddev.com/android-keystore-what-is-the-difference-between-strongbox-and-hardware-backed-keys-4c276ea78fd0).

Reply:
I think that we can add a configuration to enforce hardware backed keys. I'm already getting information about the just created key for the logging purposes, so we can utilize this and reject non-hardware backed keys if application developer wants this level of security.

Reply:
Unfortunately, such information is available only after the key is created. I don't think that I can get that status before the key is created :(

Reply:
OK - makes sense...